grafast/grafserv/src/servers/h3/v1/index.ts

```diff
@@ -242,58 +250,49 @@ export class H3Grafserv extends GrafservBase {
     }
   }
 
-  async getUpgradeHandler_experimental() {
-    if (this.resolvedPreset.grafserv?.websockets) {
-      return makeNodeUpgradeHandler(this);
-    } else {
-      return null;
+  public makeWsHandler(): Partial<Hooks> {
+    const graphqlWsServer = makeServer(makeGraphQLWSConfig(this));
+    interface Client {
+      handleMessage?: (data: string) => Promise<void>;
+      closed?: (code?: number, reason?: string) => Promise<void>;
     }
-  }
 
-  shouldHandleUpgrade_experimental(
-    req: IncomingMessage,
-    _socket: Duplex,
-    _head: Buffer,
-  ) {
-    const fullUrl = req.url;
-    if (!fullUrl) {
-      return false;
-    }
-    const q = fullUrl.indexOf("?");
-    const url = q >= 0 ? fullUrl.substring(0, q) : fullUrl;
-    const graphqlPath = this.dynamicOptions.graphqlPath;
-    return url === graphqlPath;
-  }
-
-  public async addTo_experimental(
-    app: App,
-    server: HTTPServer | HTTPSServer | undefined,
-    addExclusiveWebsocketHandler = true,
-  ) {
-    this.addTo(app);
-
-    if (addExclusiveWebsocketHandler && server) {
-      await this.attachWebsocketsToServer_experimental(server);
-    }
-  }
-
-  public async attachWebsocketsToServer_experimental(
-    server: HTTPServer | HTTPSServer,
-  ) {
-    const grafservUpgradeHandler = await this.getUpgradeHandler_experimental();
-    if (grafservUpgradeHandler) {
-      const upgrade = (req: IncomingMessage, socket: Duplex, head: Buffer) => {
-        if (this.shouldHandleUpgrade_experimental(req, socket, head)) {
-          grafservUpgradeHandler(req, socket, head);
-        } else {
-          socket.destroy();
-        }
-      };
-      server.on("upgrade", upgrade);
-      this.onRelease(() => {
-        server.off("upgrade", upgrade);
-      });
-    }
+    const clients = new WeakMap<Peer, Client>();
+    return {
+      open(peer) {
+        const client: Client = {};
+        clients.set(peer, client);
+        const onClose = graphqlWsServer.opened(
+          {
+            protocol: peer.websocket.protocol ?? GRAPHQL_TRANSPORT_WS_PROTOCOL, // will be validated
+            send(data) {
+              peer.send(data);
+            },
+            close(code, reason) {
+              peer.close(code, reason); // there are protocol standard closures
+            },
+            onMessage(cb) {
+              client.handleMessage = cb;
+            },
+          },
+          { socket: peer.websocket, request: peer.request },
+        );
+        client.closed = async (code, reason) => {
+          // @ts-expect-error fixed in unreleased https://github.com/enisdenjo/graphql-ws/pull/573
+          onClose(code, reason);
+        };
+      },
+      message(peer, message) {
+        clients.get(peer)?.handleMessage?.(message.text());
+      },
+      close(peer, details) {
+        clients.get(peer)?.closed?.(details.code, details.reason);
+        clients.delete(peer);
+      },
+      error(peer, error) {
+        // clients.delete(peer)?
+      },
```

Comment:
When an `error` occurs, does `close` also get called? I noticed that `clients` is a `Map` but it looked like it grew indefinitely so I added `clients.delete(peer)` to the `close()` event so that we wouldn't leak memory.

Reply:
clients is a WeakMap, normally, the garbage collector do it's job later, but we can delete peers manually for sure.
Thus, you're right, we need to delete peer's reference too on error.

Reply:
Oh I didn't notice it was a WeakMap. I'd much prefer it was a `Map` and we cleaned up properly than relying on `WeakMap` which can result in spiky memory usage.

Reply:
I will make these changes